Ensure proper case-less handling of FlagName

peti opened the pull request:
This patch moves the conversion to all-lower-case out of the `FlagName` parser instances and into the `mkFlagName` constructor function. 

Fixes https://github.com/haskell/cabal/issues/5043.

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

23Skidoo commented:
There are some legit CI failures: it fails to build on older GHCs and AppVeyor found this:

```
Unit Tests
  Distribution.Types.GenericPackageDescription
    findDuplicateFlagAssignments:                                                         FAIL
      tests\UnitTests\Distribution\Types\GenericPackageDescription.hs:53:
      unexpected: []
    FlagName is case-insensitive:                                                         FAIL
      *** Failed! Falsifiable (after 26 tests):
      [(FlagName "B",True),(FlagName "Vp",True),(FlagName "T",False),(FlagName "Y",False),(FlagName "u",True),(FlagName "We",False),(FlagName "I",True),(FlagName "Y",True),(FlagName "m",False),(FlagName "o",False),(FlagName "D",False),(FlagName "C",False),(FlagName "f",False),(FlagName "d",True),(FlagName "Tv",False),(FlagName "l",True),(FlagName "dM",False),(FlagName "g",False),(FlagName "sc",False),(FlagName "M",False),(FlagName "DS",False),(FlagName "aE",True),(FlagName "ea",True),(FlagName "W",False),(FlagName "df",False)]
      Use --quickcheck-replay=474123 to reproduce.
```

phadej commented on Cabal/tests/UnitTests/Distribution/Types/GenericPackageDescription.hs:
```diff
@@ -35,3 +41,33 @@ throwsUndefined :: NFData a => String -> a -> Assertion
+    let fa'        = nub [ map toLower (unFlagName n) | (n,_) <- fa ]
+        duplicates = findDuplicateFlagAssignments (mkFlagAssignment fa)
+    in
+       length duplicates == length fa - length fa'
```
use `===` (triple equals) to print LHS and RHS

peti replied:
@phadej, I force-pushed a new version (72e4f5575243305b3524d5666f5660c3f965b9c3) that uses `===`.

peti commented:
@23Skidoo, yes, of course! That failure documents the fact that our current API does not ensure that flag name are treated in a case-insensitive way. We apply `lowercase` when parsing a FlagAssignment from String, but any other way of creating that type bypasses this conversion and then `foo` and `FOO` are, in fact, two distinct flag names. These tests should succeed, but they don't.

23Skidoo commented:
Do you want to mark it as expected failure (to be fixed later) and merge as is, or do you plan to fix the underlying issue as part of the same PR?

peti commented:
@23Skidoo, I would like to fix the issue in this PR. The obvious solution would be to move the conversion to all lower-case into the `mkFlagName` function. I suggested that in https://github.com/haskell/cabal/issues/5043. So far, nobody has spoken out against that change, but curiously enough no-one has spoken out in favor of it either, so I haven't made change yet. I probably should.

23Skidoo commented:
Noted.

peti commented:
OK, I fixed the underlying issue and moved the `map toLower` conversion out of the `FlagName` parser code and into the `mkFlagName` constructor function. I also fixed the build errors with older compilers and re-based my patch to the current `master`.

There are still CI failures in Travis et al, but these failures seem to be unrelated to my changes.

peti commented:
All CI tests on travis are green. I think this PR is ready to be merged.

23Skidoo commented:
There are some failing tests in downstream CI builds.

peti commented:
I don't what is causing these CI test failures in appveyor. I have run the cabal test suite locally and it succeeds just fine on my machine.

hvr commented:
Well, there's still the issue that this PR introduces a technical debt we'll have to revert when cleaning up the semantics as e.g. suggested in https://github.com/haskell/cabal/issues/5043#issuecomment-361914463, as then we can't just normalise in `mkFlagName`; instead we'll have to move it out again, effectively reverting large parts of this PR

peti commented:
Well, there is also the issue that `mkFlagName` allows users to create a `FlagName` that violates the semantics of that type, and that problem exists *today*, whereas those new-and-improved semantics are going to exist at some point in the future of Cabal 3.0, maybe.

peti commented:
So ... I trust this all means that this PR will be lying here forever and that I might just as well close it?
